packages/core/src/render3/node_manipulation.ts

```diff
@@ -863,3 +731,158 @@ function appendProjectedNode(
+  while (viewRootTNode !== null) {
+    executeActionOnNode(renderer, action, lView, viewRootTNode, renderParent, beforeNode);
+    viewRootTNode = viewRootTNode.next;
+  }
+}
+
+/**
+ * `executeActionOnProjection` performs operation on the projection specified by `action` (insert,
+ * detach,
+ * destroy)
+ *
+ * Inserting a projection requires us to locate the projected nodes from the parent component. The
+ * complication is that those nodes themselves could be re-projected from its parent component.
+ *
+ * @param renderer Renderer to use
+ * @param action action to perform (insert, detach, destroy)
+ * @param lView The LView which needs to be inserted, detached, destroyed.
+ * @param renderParent parent DOM element for insertion/removal.
+ * @param beforeNode Before which node the insertions should happen.
+ */
+function executeActionOnProjection(
+    renderer: Renderer3, action: WalkTNodeTreeAction, lView: LView,
+    tProjectionNode: TProjectionNode, renderParent: RElement | null,
+    beforeNode: RNode | null | undefined) {
+  const componentLView = findComponentView(lView);
+  const componentNode = componentLView[T_HOST] as TElementNode;
+  const nodeToProject = componentNode.projection ![tProjectionNode.projection] !;
+  if (Array.isArray(nodeToProject)) {
+    for (let i = 0; i < nodeToProject.length; i++) {
+      const rNode = nodeToProject[i];
+      ngDevMode && assertDomNode(rNode);
+      executeActionOnElementOrContainer(action, renderer, renderParent, rNode, beforeNode);
+    }
+  } else {
+    let projectionTNode: TNode|null = nodeToProject;
+    const projectedComponentLView = componentLView[PARENT] as LView;
+    while (projectionTNode !== null) {
+      executeActionOnNode(
+          renderer, action, projectedComponentLView, projectionTNode, renderParent, beforeNode);
+      projectionTNode = projectionTNode.projectionNext;
+    }
+  }
+}
+
+
+/**
+ * `executeActionOnContainer` performs operation on the container and its views as specified by
+ * `action`
+ * (insert, detach, destroy)
+ *
+ * Inserting a Container is complicated by the fact that the container may have Views which
+ * themselves have containers or projections.
+ *
+ * @param renderer Renderer to use
+ * @param action action to perform (insert, detach, destroy)
+ * @param lContainer The LContainer which needs to be inserted, detached, destroyed.
+ * @param renderParent parent DOM element for insertion/removal.
+ * @param beforeNode Before which node the insertions should happen.
+ */
+function executeActionOnContainer(
+    renderer: Renderer3, action: WalkTNodeTreeAction, lContainer: LContainer,
+    renderParent: RElement | null, beforeNode: RNode | null | undefined) {
+  ngDevMode && assertLContainer(lContainer);
+  const anchor = lContainer[NATIVE];  // LContainer has its own before node.
+  const native = unwrapRNode(lContainer);
+  // A LContainer can be created dynamically on any node by injecting ViewContainerRef.
```

Comment:
Thank you for this comment !

```suggestion
  // An LContainer can be created dynamically on any node by injecting ViewContainerRef.
```